bftengine/src/bftengine/KeyManager.h

```diff
@@ -13,49 +13,124 @@
 
 #include "InternalBFTClient.h"
 #include "KeyStore.h"
+#include "bftengine/MultiSignatureInterfaces.hpp"
+#include "Timers.hpp"
+#include "Metrics.hpp"
+#include "SequenceWithActiveWindow.hpp"
+#include "SeqNumInfo.hpp"
 
 namespace bftEngine::impl {
 class KeyManager {
  public:
-  static KeyManager& get(InternalBFTClient* cl = nullptr,
-                         const int id = 0,
-                         const uint32_t clusterSize = 0,
-                         IReservedPages* reservedPages = nullptr,
-                         const uint32_t sizeOfReservedPage = 0) {
-    static KeyManager km{cl, id, clusterSize, reservedPages, sizeOfReservedPage};
-    return km;
-  }
-
+  // Generates and publish key to consensus
   void sendKeyExchange();
+  // Generates and publish the first key
+  void sendInitialKey();
+  // The execution handler implementation that is called when a key exchange msg has passed consensus.
   std::string onKeyExchange(KeyExchangeMsg& kemsg, const uint64_t& sn);
+  // A callback that is called each checkpoint and can trigger rotation of keys.
   void onCheckpoint(const int& num);
+  // Register a IKeyExchanger to notification whehn keys are rotated.
   void registerForNotification(IKeyExchanger* ke);
   KeyExchangeMsg getReplicaKey(const uint16_t& repID) const;
   void loadKeysFromReservedPages();
 
+  // loads the crypto system from a serialized key view.
   std::atomic_bool keysExchanged{false};
+  std::future<void> futureRet;
+
+  struct InitData {
+    InternalBFTClient* cl{nullptr};
+    int id{};
+    uint32_t clusterSize{};
+    IReservedPages* reservedPages{nullptr};
+    uint32_t sizeOfReservedPage{};
+    IPathDetector* pathDetect{nullptr};
+    concordUtil::Timers* timers{nullptr};
+    std::shared_ptr<concordMetrics::Aggregator> a;
+    std::chrono::seconds interval;
+  };
+
+  static KeyManager& get(InitData* id = nullptr) {
+    static KeyManager km{
+        id->cl, id->id, id->clusterSize, id->reservedPages, id->sizeOfReservedPage, id->pathDetect, *(id->timers)};
+    return km;
+  }
+
+  static void start(InitData* id) {
+    get(id);
+    get().initMetrics(id->a, id->interval);
+  }
 
  private:
-  KeyManager(InternalBFTClient* cl,
+  KeyManager(IinternalBFTClient* cl,
```

Comment:
Why not to use InitData as a parameter to the constructor also?

Reply:
ok